### PR TITLE
Update ECIES implementation for crypto 3.2

### PR DESCRIPTION
**Pull request by romanstrobl**
This pull request introduces updated ECIES implementation for crypto 3.2.

### CODE REVIEW

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/PowerAuthService.java [hunk 1]
```diff
@@ -1156,6 +1175,31 @@ public GetEciesDecryptorResponse getEciesDecryptor(GetEciesDecryptorRequest requ
...
+        // The activationId value can be null in case the decryptor is used in application scope
+        try {
+            logger.info("GetEciesEncryptorRequest received, application key: {}, activation ID: {}", request.getApplicationKey(), request.getActivationId());
+            final GetEciesEncryptorResponse response = behavior.getEciesEncryptionBehavior().getEciesEncryptorParameters(request);
+            logger.info("GetEciesEncryptorRequest succeeded");
+            return response;
+        } catch (GenericServiceException ex) {
+            // already logged
+            throw ex;
+        } catch (RuntimeException | Error ex) {
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/PowerAuthService.java:1194]: We should not catch `Error`. The fix will be done in a different PR but I suggest not to copy&paste this pattern to the new code. See https://rules.sonarsource.com/java/RSPEC-1181/

**Review by banterCZ — APPROVED** [review]
Looks good.

File: powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/request/GetEciesDecryptorRequest.java [hunk 2]
```diff
@@ -31,5 +31,8 @@ public class GetEciesDecryptorRequest {
     private String activationId;
     private String applicationKey;
     private String ephemeralPublicKey;
+    private String nonce;
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/request/GetEciesDecryptorRequest.java:34]: Sure, we can.

File: powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/request/GetEciesEncryptorRequest.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * @author Roman Strobl, roman.strobl@wultra.com
+ */
+@Data
+public class GetEciesEncryptorRequest {
+
+    private String activationId;
+    private String applicationKey;
+    private String ephemeralPublicKey;
+    private String nonce;
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/request/GetEciesEncryptorRequest.java:34]: Sure, we can.

File: powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/response/CreateActivationResponse.java [hunk 4]
```diff
@@ -34,6 +34,9 @@ public class CreateActivationResponse {
     private String applicationId;
     private String encryptedData;
     private String mac;
+    private String ephemeralPublicKey;
+    private String nonce;
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/response/CreateActivationResponse.java:38]: Ok.

File: powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/response/CreateTokenResponse.java [hunk 5]
```diff
@@ -30,5 +30,8 @@ public class CreateTokenResponse {
 
     private String encryptedData;
     private String mac;
+    private String ephemeralPublicKey;
+    private String nonce;
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/response/CreateTokenResponse.java:34]: Ok.

File: powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/response/PrepareActivationResponse.java [hunk 6]
```diff
@@ -34,6 +34,9 @@ public class PrepareActivationResponse {
     private String applicationId;
     private String encryptedData;
     private String mac;
+    private String ephemeralPublicKey;
+    private String nonce;
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/response/PrepareActivationResponse.java:38]: Ok.

File: powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/response/RecoveryCodeActivationResponse.java [hunk 7]
```diff
@@ -34,6 +34,9 @@ public class RecoveryCodeActivationResponse {
     private String applicationId;
     private String encryptedData;
     private String mac;
+    private String ephemeralPublicKey;
+    private String nonce;
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/response/RecoveryCodeActivationResponse.java:38]: Ok.

File: powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/response/StartUpgradeResponse.java [hunk 8]
```diff
@@ -30,5 +30,8 @@ public class StartUpgradeResponse {
 
     private String encryptedData;
     private String mac;
+    private String ephemeralPublicKey;
+    private String nonce;
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-client-model/src/main/java/com/wultra/security/powerauth/client/model/response/StartUpgradeResponse.java:34]: Ok.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/PowerAuthService.java (hunk 1 above)
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/PowerAuthService.java:1194]: Let's raise a separate issue.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/PowerAuthServiceConfiguration.java [hunk 9]
```diff
@@ -155,6 +155,13 @@ public class PowerAuthServiceConfiguration {
     @Max(8)
     private int offlineSignatureComponentLength;
 
+    /**
+     * Expiration of timestamps for ECIES and MAC token requests.
+     */
+    @Value("${powerauth.service.crypto.requestExpirationInMilliseconds}")
+    @Min(0)
+    private int requestExpirationInMilliseconds;
```
Comment (banterCZ) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/PowerAuthServiceConfiguration.java:163]: Just an idea, may we use `java.time.Duration` instead?

File: docs/db/changelog/changesets/powerauth-java-server/1.5.x/20230723-add-table-unique-value.xml [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+    <!--  Create table pa_unique_value -->
+    <changeSet id="1" logicalFilePath="powerauth-java-server/1.5.x/20230723-add-table-unique-value.xml" author="Roman Strobl">
+        <preConditions onFail="MARK_RAN">
+            <not>
+                <tableExists tableName="pa_unique_value"/>
+            </not>
+        </preConditions>
+        <comment>Create a new table pa_unique_value</comment>
+        <createTable tableName="pa_unique_value">
```
Comment (banterCZ) [inline docs/db/changelog/changesets/powerauth-java-server/1.5.x/20230723-add-table-unique-value.xml:14]: Please describe this change in the migration guide as well.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/PowerAuthServiceConfiguration.java (hunk 9 above)
Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/PowerAuthServiceConfiguration.java:163]: Let's raise a separate issue, because we will need to refactor other usages as well.

File: docs/db/changelog/changesets/powerauth-java-server/1.5.x/20230723-add-table-unique-value.xml (hunk 10 above)
Comment (romanstrobl) [reply docs/db/changelog/changesets/powerauth-java-server/1.5.x/20230723-add-table-unique-value.xml:14]: Good point, I will add it.

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/PowerAuthServiceConfiguration.java (hunk 9 above)
Comment (banterCZ) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/configuration/PowerAuthServiceConfiguration.java:163]: Ok, I have created #940

File: powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/util/EciesDataUtils.java [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        if ("3.2".equals(version)) {
+            if (eciesScope == EciesScope.ACTIVATION_SCOPE) {
+                return ByteUtils.concatStrings(version, applicationKey, activationId);
+            } else {
+                return ByteUtils.concatStrings(version, applicationKey);
+            }
+        } else {
+            return null;
+        }
+    }
```
Comment (hvge) [inline powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/util/EciesDataUtils.java:48]: I think this function should be implemented in `powerauth-crypto`. I wonder how we do this in cmd-lib and cmd-tool.

Comment (hvge) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/util/EciesDataUtils.java:48]: I actually miss it there, because I wanted to test the `powerauth-crypto` ECIES functionality against my test vectors but there's no such function available in the crypto library.

Comment (romanstrobl) [reply powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/util/EciesDataUtils.java:48]: Agree, I will move it.

File: docs/WebServices-Methods.md [hunk 12]
```diff
@@ -1460,6 +1461,31 @@ REST endpoint: `POST /rest/v3/ecies/decryptor`
...
+
+REST endpoint: `POST /rest/v3/ecies/encryptor`
+
+`GetEciesEncryptorRequest`
+
+| Type | Name | Description |
+|------|------|-------------|
+| `String` | `activationId` | A UUID4 identifier of an activation (used only in activation scope, use null value in application scope) |
+| `String` | `applicationKey` | A key (identifier) of an application, associated with given application version |
+| `String` | `ephemeralPublicKey` | A base64 encoded ephemeral public key for ECIES |
```
Comment (hvge) [inline docs/WebServices-Methods.md:1478]: Do we have use-case for this? In `powerauth-restful-integration` project we don't use this API at all so I wonder why we need such API.

If it provide the same data as `/ecies/decryptor` then let's remove it as redundant.

Comment (romanstrobl) [reply docs/WebServices-Methods.md:1478]: You are right, we can use the original endpoint. I have not realized the returned data is the same.
